lib/server.js

```diff
@@ -0,0 +1,205 @@
+    bytes.push(parseInt(hex.substr(i, 2), 16))
+  }
+  return Buffer.from(bytes)
+}
+
+/**
+ * Filter incoming headers based on allowed headers.
+ * We always want to send Pascal-Case headers so we use the key from constants
+ * with the value from the incoming header.
+ * @param {Map<string, boolean>} allowedHeaders
+ * @returns {(incomingHeaders: Headers) => Record<string, string>}
+ * */
+function filterHeaders(allowedHeaders) {
+  return function (incomingHeaders) {
+    /** @type {Record<string, string>} */
+    const filteredHeaders = {}
+
+    for (const [key, bool] of allowedHeaders.entries()) {
+      if (incomingHeaders.has(key.toLowerCase()) && bool) {
+        // @ts-ignore
+        filteredHeaders[key] = incomingHeaders.get(key.toLowerCase())
+      }
+    }
+
+    return filteredHeaders
+  }
+}
+
+/*
+ * Node.js HTTP server to proxy insecure content via HTTPS.
+ */
+export class Server extends EventEmitter {
+  /**
+   * @param {Config} config
+   */
+  constructor(config) {
+    super()
+    if (!config.HMACKey) {
+      throw new Error('HMACKey is required')
+    }
+    this.config = config
+    this.config.serverName ??= 'camo'
+  }
+
+  /**
+   * Start the server. Identical to `net.Server.listen()`.
+   * @param {Parameters<InstanceType<typeof net.Server>['listen']>} args
+   * @public
+   */
+  listen(...args) {
+    return http.createServer(this.handle.bind(this)).listen(...args)
+  }
+
+  /**
+   * The main request handler handles all incoming requests.
+   * Integrate with your own server by calling this method and routing all requests to it.
+   * @param {http.IncomingMessage} req
+   * @param {http.ServerResponse} res
+   * @public
+   */
+  async handle(req, res) {
+    if (req.method !== 'GET' && req.method !== 'HEAD') {
+      return this.write(res, 405, 'Method not allowed')
+    }
+
+    // @ts-ignore req.url is fine
+    const paths = url.parse(req.url)?.path?.split('/')
+
+    if (!paths || paths.length < 3) {
+      return this.write(res, 404, 'Malformed request')
+    }
+
+    try {
+      const [, receivedDigest, hex] = paths
+      var decodedUrl = this.verifyHMAC(receivedDigest, hex)
+    } catch (err) {
+      return this.write(res, 403, 'Bad signature')
+    }
+
+    try {
+      var validUrl = await SafeHttpClient.checkUrl(decodedUrl)
+    } catch (err) {
+      // @ts-ignore err does have message
+      return this.write(res, 400, err.message)
+    }
+
+    const controller = new AbortController()
+    const {signal} = controller
+
+    req.on('close', () => {
+      controller.abort()
+    })
+
+    try {
+      // TODO: respect forwarded headers (check if not private IP)
+      const filterRequestHeaders = filterHeaders(REQUEST_HEADERS)
+      const filterResponseHeaders = filterHeaders(RESPONSE_HEADERS)
+      const client = new SafeHttpClient(this.config.maxSize)
+      const {buffer, headers: resHeaders} = await client.safeFetch(validUrl, {
+        // @ts-ignore we can convert req.headers to Headers
```

Comment:
```suggestion
        // @ts-expect-error: `IncomingHttpHeaders` can be passed to `Headers`
```

a) use ts-expect-error, which results in an error if it doesn’t error, that way, when Node types are fixed, we get an error, and can remove it